### ISSUE TITLE
Why clipping is not as bad as it looks, and how we can use the features to make recommendations for better map-design. 

### ISSUE THREAD

**Issue by Nytilde** [PKBeam/Edda#120]
The features are used in the new difficulty predictor: ["AverageTimeDifference", "d1min", "HighNoteDensity2s", "NoteDensity", "cdw"]

To analyze the behavior of a regular map, one must initially ascertain its rate of change. This is accomplished by differentiating a time function with respect to time. However, in practical applications, we utilize a time vector to compute time differences, enabling the determination of the iterator's change with a lag of one:

Figure 1: Rate of Change
![image](https://github.com/PKBeam/Edda/assets/91031776/f4ae700a-f0de-4ba2-9f35-2162b0fde96e)

The presence of high peaks in the data indicates longer breaks, while the lower sections indicate a higher density of notes in the area. Ultimately, we use the average rate of change, which heuristically provides the best general separation performance according to the Exploratory Data Analysis (EDA). This separation performance in the level 1 range could be further improved with the 4th quantile.

HighNoteDensity2s corresponds approximately to the rolling density and has a high classification performance in the upper level range but decreases in the lower level range.

Figure 2: HighNoteDensity2s
![image](https://github.com/PKBeam/Edda/assets/91031776/38a5acaf-5b2e-4aee-8faa-7c9476c152c5)

NoteDensity offers a generally uniform but not so pronounced differentiation of all level ranges.

Figure 3: NoteDensity
![image](https://github.com/PKBeam/Edda/assets/91031776/dba0927e-b6be-4355-b71e-a9b0570ac768)

cdw follows the approach of counting all notes in a constant window range in the higher velocity range. This feature contributes only minimally to the improvement but could slightly raise the classification in the middle range.

Figure 4: cdw
![image](https://github.com/PKBeam/Edda/assets/91031776/2b5b30fb-63f5-4b92-a155-a3e4a4538d8b)

The next step is to introduce the influence of the pattern (work in progress).

What do models that are filtered via clipping tell us? In principle, we need to look in detail at which features do not match, and we can determine from the resulting deviations.

According to my observations, most outliers in the lower and middle levels often indicate little variety in the beatmap.
Recommendation: include more breaks to make it more varied.

I am always open to further suggestions and constructive criticism of my ideas.

**Comment by Nytilde** [comment]
Here is the link to the Live Playground:
https://colab.research.google.com/drive/1hj86xf_RDBks0uO6CDi4olxMlwilHVQC?usp=sharing

**Comment by Nytilde** [comment]
A further idea could be a fall-back model, which, while having lower accuracy, is designed specifically for extrapolations. To explore this, I have conducted a test where I reduced the dimensions of the PCA to 1 and 2:

Figure 1: Current Model
![current model](https://github.com/PKBeam/Edda/assets/91031776/22359086-abf1-4f58-a40d-0cf535612cac)

Figure 2: 1-dimensional
![1dim](https://github.com/PKBeam/Edda/assets/91031776/3bfc9df9-21c2-4fcc-9ec7-de04d00ce79d)

Figure 3: 2-dimensional
![2dim](https://github.com/PKBeam/Edda/assets/91031776/456803d6-216f-4ccd-ba1a-37874941feea)

According to the theory, this approach should work, although I haven't tested it yet.

## My Considerations
The disadvantage of this method is that during extrapolation, only the influence of the signal size is taken into account, not the combination of factors. This results in an extrapolated mean that is generated with an asymmetrical map design. For example, this might mean having level 1 in the first half and level 10 in the second half.

**Comment by Brollyy** [comment]
Idea from Mrnaris on Discord:
> Would be cool if there was a way to "check" the difficulty within an interval, for example, if i wanna know how hard the hardest part of a song is i'd select that part and it would give me a difficulty value for that part, same goes for if i wanna see how hard the easiest part of a song is, idk how hard this would be to implement tho.

> the interval could possibly be set by two different bookmarks

**Comment by Brollyy** [comment]
![image](https://github.com/user-attachments/assets/fd4f4749-f204-4736-9891-f11cf2ae4800)
![image](https://github.com/user-attachments/assets/65e4f377-a351-4aa3-9967-34290a60f529)

# What's next?

There's been a lot of feedback regarding the new difficulty predictor tool, we should consider which way we want to go with it then. @PKBeam @Nytilde

I'm all for allowing users to select whichever model/algorithm they want to use while we work on improving the models, but I have a few points that I think we'll need to look at first:
1. I'm afraid that the differences between models will be hard to understand, so it'll be confusing and not very helpful to see very different difficulty estimations from different models. We could add descriptions to the options, but that will make the Difficulty Predictor window UI even more busier than it already is...
2. How would the model selection work with #119 ? From what I understood, the old model is best when used with fully-completed map, so it wouldn't make sense to display real-time difficulty prediction for that option.
3. If we go with multiple options for the user, do we include the algorithm suggested by Melchior on Discord? From what I understood, Nytilde used it as an additional feature for the ML model and saw some improvement on higher difficulties, but we could also put it as a separate option altogether - here's the results of fitting it to all OST and RAID songs: https://docs.google.com/spreadsheets/d/12osrf0tNd8vZQqO2ahS8AMeypqug8wgg8gZsEJnuSwo/edit?usp=sharing
4. Should we allow the models to not provide any estimation if the map params are outside the expected range (like with ">10" right now)? I think that adds to the confusion right now - it's probably better to just have a fallback model Nytilde described earlier for this case.
5. Gathering feedback on the difficulty estimations in-app (e.g. thumbs up/thumbs down) would be very helpful for future improvements, but I think we'd need to host a new server to store it, so probably too much of a hassle...

In the meantime, I'll work at least on implementing the new version of Nytilde's model + the fallback model that she sent me 2 months back to see if there are any improvements.

**Comment by Brollyy** [comment]
The changes done in #129 are a good step forward, but I'm not closing this issue, as there are still unresolved problems with the new ML model that need addressing in the future releases.

**Comment by Nytilde** [comment]
I have been reviewing the data that all of you have provided to me.
The unexpected behavior might be due to multicollinearity, where some features are highly correlated with each other. 
To address this, instead of using a model that might be overly sensitive to such deviations, we should consider using Multiple Linear Regression (MLR).  

![Linear Regression Formula](https://latex.codecogs.com/png.latex?\bg_white\dpi{150}\hat{y}=w_0+w_1x_1+w_2x_2+\ldots+w_nx_n)

This approach models the target variable as a linear function of the features. So it should be a more stable and interpretable solution.

TLDR: To model a linear behavior, it is necessary to describe it as a linear function.

I don't have that much time those days to do that.
Perhaps it is better if we train a simpler model for now, which uses a maximum of 3 dimensions.
